src/Microsoft.AspNetCore.HostFiltering/HostFilteringMiddleware.cs

```diff
@@ -0,0 +1,182 @@
+            + "<hr><p>HTTP Error 400. The request hostname is invalid.</p>\r\n"
+            + "</BODY></HTML>");
+
+        private readonly RequestDelegate _next;
+        private readonly ILogger<HostFilteringMiddleware> _logger;
+        private readonly HostFilteringOptions _options;
+        private readonly IServerAddressesFeature _serverAddresses;
+        private IList<StringSegment> _allowedHosts;
+        private bool? _allowAnyNonEmptyHost;
+
+        /// <summary>
+        /// A middleware used to filter requests by their Host header.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="serverAddresses"></param>
+        /// <param name="logger"></param>
+        /// <param name="options"></param>
+        public HostFilteringMiddleware(RequestDelegate next, ILogger<HostFilteringMiddleware> logger, 
+            IOptions<HostFilteringOptions> options, IServerAddressesFeature serverAddresses)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+            _serverAddresses = serverAddresses ?? throw new ArgumentNullException(nameof(serverAddresses));
+        }
+
+        /// <summary>
+        /// Processes requests
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public Task Invoke(HttpContext context)
+        {
+            EnsureConfigured();
+
+            if (!CheckHost(context))
+            {
+                context.Response.StatusCode = 400;
+                if (_options.IncludeFailureMessage)
+                {
+                    context.Response.ContentLength = DefaultResponse.Length;
+                    context.Response.ContentType = "text/html";
+                    return context.Response.Body.WriteAsync(DefaultResponse, 0, DefaultResponse.Length);
+                }
+                return Task.CompletedTask;
+            }
+
+            return _next(context);
+        }
+
+        private void EnsureConfigured()
+        {
+            if (_allowAnyNonEmptyHost == true || _allowedHosts?.Count > 0)
+            {
+                return;
+            }
+
+            var allowedHosts = new List<StringSegment>();
+            if (_options.AllowedHosts?.Count > 0)
+            {
+                if (!TryProcessHosts(_options.AllowedHosts, allowedHosts))
+                {
+                    _logger.LogDebug("Wildcard detected, all requests with hosts will be allowed.");
+                    _allowAnyNonEmptyHost = true;
+                    return;
+                }
+            }
+            else
+            {
+                if (!TryProcessHosts(_serverAddresses.Addresses.Select(address => BindingAddress.Parse(address).Host), allowedHosts))
```

Comment:
@shirhatti @davidfowl thinking about this more I realized that the use of IServerAddresses here is nonsensical. You need the middleware when there's something deficient about the server configuration, so the middleware will always need different values than the server. E.g. for the primary use case Kestrel, even if you specify https://foo.com:443/ it actually populates IServerAddresses with https://[::]:443/ which is useless to us. Copying config from ANCM in proc or HttpSys is also useless because you'd only be repeating the same validation they'd already done (or none if they were set to wildcards). For the in-proc scenarios you need correctly configure the server and exclude this middleware or set it to * to disable it.

I'm going to remove IServerAddresses from this middleware.